### PR TITLE
Merge all of the room membership codepaths into one

### PR DESCRIPTION
**Pull request by illicitonion**
Previously, there were several codepaths all of which performed inconsistent auth checks, and had different ways of looking up remote room IDs, inviters, and such. Now there is one.

It also fixes several minor bugs along the way.

### CODE REVIEW

File: synapse/handlers/message.py [hunk 1]
```diff
@@ -226,55 +226,69 @@ def send_event(self, event, context, ratelimit=True, is_guest=False):
             ratelimit (bool): Whether to rate limit this send.
             is_guest (bool): Whether the sender is a guest.
         """
+        if event.type == EventTypes.Member:
+            raise SynapseError(
+                500,
+                "Tried to send member even through non-member codepath"
```
Comment (NegativeMjark) [inline synapse/handlers/message.py]: "event"

File: synapse/handlers/room.py [hunk 2]
```diff
@@ -432,202 +442,182 @@ def update_membership(self, requester, target, room_id, action, txn_id=None):
...
-                        event.user_id
-                    )
-                    defer.returnValue({"room_id": room_id})
-                    return
-
-            # FIXME: This isn't idempotency.
-            if prev_state and prev_state.membership == event.membership:
-                # double same action, treat this event as a NOOP.
-                defer.returnValue({})
+        if event.is_state():
```
Comment (NegativeMjark) [inline synapse/handlers/room.py]: When would a membership event not be a state event?

File: synapse/handlers/message.py (hunk 1 above)
Comment (illicitonion) [inline synapse/handlers/message.py]: Done

File: synapse/handlers/room.py (hunk 2 above)
Comment (illicitonion) [inline synapse/handlers/room.py]: Yay for copy and paste!

Done

**Comment by NegativeMjark** [comment]
LGTM

**Pull request merged**